pkg/cmd/codespace/ssh.go

```diff
@@ -298,6 +306,118 @@ func checkAndUpdateOldKeyPair(sshContext ssh.Context) *ssh.KeyPair {
 	return nil
 }
 
+// hasUploadedValidCustomKey checks which private keys are used in ssh configration and
+// compares them to uploaded public keys to see if there is a match
+func hasUploadedPublicKeyForConfig(
+	ctx context.Context,
+	sshContext ssh.Context,
+	apiClient apiClient,
+	customConfigFile string,
+	customHost string,
+) bool {
+	configuredPrivateKeyPaths, err := getConfiguredPrivateKeys(ctx, customConfigFile, customHost)
+	if err != nil {
+		return false
+	}
+
+	configuredPublicKeys := []string{}
+	for _, privateKeyPath := range configuredPrivateKeyPaths {
+		publicKeyPath := privateKeyPath + ".pub"
+
+		publicKeyContent, err := os.ReadFile(publicKeyPath)
+		if err != nil {
+			// The default configuration includes standard keys like id_rsa,
+			// id_ed25519, etc, but these may not actually exist so just skip
+			continue
+		}
+
+		configuredPublicKeys = append(configuredPublicKeys, string(publicKeyContent))
+	}
+
+	if len(configuredPublicKeys) == 0 {
+		// There are no local private keys which ssh would use
+		return false
+	}
+
+	user, err := apiClient.GetUser(ctx)
+	if err != nil {
+		return false
+	}
+
+	uploadedPublicKeys, err := apiClient.AuthorizedKeys(ctx, user.Login)
+	if err != nil {
+		return false
+	}
+
+	if len(uploadedPublicKeys) == 0 {
+		return false
+	}
+
+	for _, configuredPublicKey := range configuredPublicKeys {
+		for _, uploadedPublicKey := range uploadedPublicKeys {
```

Comment:
This is probably not a problem, but it may be worth hashing `uploadedPublicKeys` into a `map[string]bool` and do O(1) lookups vs re-iterating them for every local key. Personally, I have a very little amount of SSH keys, but I know of users that generate a new key for almost everything.

Reply:
Yeah, I figured that even people who generate a ton of ssh keys aren't going to run into issues here really - because we are only checking keys which would actually be used by the connection and that number should be small. Even if it was like 100 keys it would be 100x100 in the worst case which isn't bad. Anyways, the map is a quick change so I just made it :smile: